omod/src/main/java/org/openmrs/module/pihcore/fragment/controller/dashboardwidgets/DiscreteValuesFragmentController.java

```diff
@@ -0,0 +1,102 @@
+
+    public void controller(
+                           @SpringBean("conceptService") ConceptService conceptService,
+                           @SpringBean("obsService") ObsService obsService,
+                           @SpringBean("programWorkflowService") ProgramWorkflowService programWorkflowService,
+                           @FragmentParam("app") AppDescriptor app,
+                           FragmentConfiguration config,
+                           FragmentModel model) {
+
+        Object patientConfig = config.get("patient");
+        Patient patient = null;
+        if (patientConfig != null && (patientConfig instanceof PatientDomainWrapper)) {
+            patient = ((PatientDomainWrapper) patientConfig).getPatient();
+        }
+        if (patient == null) {
+            throw new IllegalArgumentException("No patient found. Please pass patient into the configuration");
+        }
+
+        model.put("app", app);
+        model.put("patient", patient);
+
+        List<SimpleObject> fieldsObject = (List<SimpleObject>) config.get("fields");
+        if (fieldsObject == null) {
+            throw new IllegalStateException("Missing configuration fields on widget");
+        }
+        Set<Concept> concepts = new HashSet<>();
+        Map<String, SimpleObject> fields = new LinkedHashMap<>();
+        for (int i = 0; i < fieldsObject.size(); i++ ) {
+            try {
+                Map<String, String> fieldNode = (LinkedHashMap) fieldsObject.get(i);
+                String label = (String) fieldNode.get("label");
+                String conceptUuid = (String) fieldNode.get("concept");
+                String minValue = (String) fieldNode.get("minValue");
+                if (StringUtils.isNotBlank(conceptUuid)) {
+                    Concept concept = conceptService.getConceptByReference(conceptUuid);
+                    if (concept == null) {
+                        throw new IllegalArgumentException("No concept with this UUID: " + conceptUuid + " found. Please pass correct concept UUID into the configuration");
+                    }
+                    concepts.add(concept);
+                    fields.put(conceptUuid, SimpleObject.create(
+                            "label", label,
+                            "minValue", minValue
+                    ));
+                }
+            } catch (Exception e) {
+                throw new IllegalStateException("Invalid configuration for encounterType node", e);
+            }
+        }
+
+        String programUuid = (String) config.get("duringCurrentEnrollmentInProgram");
+        List<Obs> obsList = PihCoreUtils.getObsWithinProgram(patient, concepts, programUuid);
+
+        for (Obs obs : obsList) {
+            obs.getValueAsString(Context.getLocale());
+            SimpleObject simpleObject = fields.get(obs.getConcept().getUuid());
+            if (simpleObject != null ) {
+                simpleObject.put("obsValue", obs.getValueAsString(Context.getLocale()));
```

Comment:
Why are we getting value as String here?  Why not just put the entire obs in here?  It's easy enough to format this as a String in the gsp by using `ui.format(obs)` (I think), and don't you need to get the numeric value for usage in the gsp to determine if it is outside of range?

Reply:
ok